### DISCUSSION TITLE
ESP32 restarting continually after HomeSpan starts up

### DISCUSSION THREAD

**Discussion by clokey** [HomeSpan/HomeSpan#830] [Q&A] [answered]
Hi,

I am trying to use homespun to implement a HomeKit enabled presence sensor built around the seeed studio [Human Static Presence Module](https://www.seeedstudio.com/24GHz-mmWave-Sensor-Human-Static-Presence-Module-Lite-p-5524.html) using an  [ESP32c3](https://www.seeedstudio.com/Seeed-XIAO-ESP32C3-p-5431.html).

All of my code is in this [Repo](https://github.com/clokey/HomeKitPresenceSensor)

I have got HomeSpan up and running without the Device specific implementation for the MR24HPC_1, I based this around the motion sensor code that was linked from the HomeSpan repo. However, when I actually include the DEV_MR24HPC_1.h code into my sketch and upload it to my device, the ESP32 goes into an endless reboot loop. I've moved the circuit around on the esp32c3 so I wasn't accidentally using strapping pins and I've tried to up the level of debugging coming from HomeSpan, but I am struggling to find some clues as to what I am doing wrong and where to go next. 

Here is an example "endless loop" having uploaded the sketch and started monitoring. It just cycles saying that HomeSpan has started and then restarting again. I never get the opportunity to configure wifi credentials, but have had this working in a simplified sketch.

Any useful pointers, much appreciated. 

```
Wrote 1247504 bytes (719105 compressed) at 0x00010000 in 9.2 seconds (effective 1089.4 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
--- Terminal on /dev/cu.usbmodem1101 | 9600 8-N-1
--- Available filters and text transformations: colorize, debug, default, direct, esp32_exception_decoder, hexlify, log2file, nocontrol, printable, send_on_enter, time
--- More details at https://bit.ly/pio-monitor-filters
--- Quit: Ctrl+C | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x3 (RTC_SW_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x403819b6
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd5810,len:0x438
load:0x403cc710,len:0x918
load:0x403ce710,len:0x24e4
entry 0x403cc710
Starting up...

************************************************************
Welcome to HomeSpan!
Apple HomeKit for the Espressif ESP-32 WROOM and Arduino IDE
************************************************************

** Please ensure serial monitor is set to transmit <newlines>

Message Logs:     Level 0
Status LED:       Pin 5
Device Control:   Pin 3
Sketch Version:   n/a
HomeSpan Version: 1.9.0
Arduino-ESP Ver.: 2.0.6
ESP-IDF Version:  4.4.3
ESP32 Chip:       ESP32-C3 Rev 4 single-core 4MB Flash
ESP32 Board:      XIAO_ESP32C3
PWM Resources:    6 channels, 4 timers, max 14-bit duty resolution
Sodium Version:   1.0.12-idf  Lib 9.4
MbedTLS Version:  mbed TLS 2.28.1
Sketch Compiled:  Mar 18 2024 17:24:36
Partition:        app0
MAC Address:      40:4C:CA:F4:D0:3C

Device Name:      Occupancy Sensor


*** AutoPolling Task started with priority=1

Ready.

*** HomeSpan Info ***

➤ Accessory:  AID=1
   ➟ Service AccessoryInformation:  IID=1, UUID="3E"
      ⇨ Characteristic Identify(1):  IID=2, UUID="14", Perms=PW
      ⇨ Characteristic Manufacturer("Clokey"):  IID=3, UUID="20", Perms=PR
      ⇨ Characteristic Model("MR24HPC1"):  IID=4, UUID="21", Perms=PR
      ⇨ Characteristic Name("Occupancy Sensor"):  IID=5, UUID="23", Perms=PR
   ➟ Service OccupancySensor:  IID=6, UUID="86"
      ⇨ Characteristic OccupancyDetected(1):  IID=7, UUID="71", Perms=PR+EV, Range=[0,1]

Service                             UUID         AID  IID  Update  Loop  Button  Linked Services
------------------------------  --------  ----------  ---  ------  ----  ------  ---------------
AccessoryInformation                  3E           1    1      NO    NO      NO  -
OccupancySensor                       86           1    6      NO   YES      NO  -

Configured as Bridge: NO

Database Validation:  Warnings=0, Errors=0

*** End Info ***

Accessory ID:      35:72:F7:F4:7B:41                               LTPK: 9AB6DF5BEC60FB3646249E9649EB26C4B648BC567F824CC24D26C102C17D41FD
No Paired Controllers

Accessory configuration number: 1

*** WIFI CREDENTIALS DATA NOT FOUND.  YOU MAY CONFIGURE BY TYPING 'W <RETURN>'.

Occupancy Sensor is READY!

ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x3 (RTC_SW_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x403819b6
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd5810,len:0x438
load:0x403cc710,len:0x918
load:0x403ce710,len:0x24e4
entry 0x403cc710
Starting up...
```

**Comment by HomeSpan** [comment]
Have you been able to get the sensor to work on the esp32 without HomeSpan (ie just standalone)?

**Reply by clokey** [reply]
Yes (should have pointed that out!), Whilst the sensor can be accessed using a somewhat convoluted serial interface I am using the simple s1, s2 pins to measure presence and activity. So the code in the device simply reads a pin to see whether it is high or low.

**Reply by HomeSpan** [reply]
What happens if you run the HomeSpan sketch configured for the sensor, but  without actually connecting the sensor to the esp32.

**Reply by clokey** [reply]
I've disconnected the power from the sensor (as well as the input wire that is read from the sketch) and I still get the same result, e.g. the endless rebooting of the esp32. I saw in another thread that you said there was an esp library command to determine reset status, would that help provide some clues?

**Reply by HomeSpan** [reply]
This suggests you may be trying to use a pin for the status LED, the control button, or the sensor, that is not applicable for the C3 chip or overlaps with some other chip function.

**Reply by clokey** [reply]
That was my initial thought, so I moved the led, control button and sensor to inputs which didn't mirror strapping pins. The main reason for asking here is that if I comment out line 28 in the main.cpp file

`  //new DEV_MR24HPC1_Sensor(OCCUPIED_PIN);`

compile and upload, everything works as expected, the esp doesn't restart and the HomeSpan cli works, and I can configure behaviour via the control button.

**Reply by HomeSpan** [reply, marked as answer]
I reviewed your code in more detail and think I spotted the problem: see line 15 of the sensor.h file - looks like you are redefining occupied as a boolean, which masks the version of occupied that points to the SpanCharacterstic.  As a result, the setVal() call in the loop() will blow up since the pointer is never instantiated.   

Perhaps this was a line you intended for debugging purposes?

**Reply by clokey** [reply]
What sharp eyes! That was indeed the issue, I was over writing the pointer with a boolean value. How I love C! 

Thanks for your help
